network/netplay/netplay_frontend.c

```diff
@@ -7191,6 +7419,99 @@ static void netplay_announce(void)
       free(frontend_ident);
 }
 
+static bool netplay_mitm_query_wait(void *data)
+{
+   bool *finished = data;
+
+   return !(*finished);
+}
+
+static void netplay_mitm_query_cb(retro_task_t *task, void *task_data,
+   void *user_data, const char *error)
+{
+   char *buf_start, *buf;
+   size_t remaining;
+   net_driver_state_t *net_st     = &networking_driver_st;
+   struct netplay_room *host_room = &net_st->host_room;
+   http_transfer_data_t *data     = task_data;
+   bool                 *finished = user_data;
+
+   if (error)
+      goto done;
+   if (!data || !data->data || !data->len)
+      goto done;
+   if (data->status != 200)
+      goto done;
+
+   buf_start = malloc(data->len);
+   if (!buf_start)
+      goto done;
+   memcpy(buf_start, data->data, data->len);
+
+   buf = buf_start;
+   remaining = data->len;
+   do
+   {
+      char *lnbreak, *delim;
+      char *key, *value;
+
+      lnbreak = (char*) memchr(buf, '\n', remaining);
+      if (!lnbreak)
+         break;
+      *lnbreak++ = '\0';
+
+      delim   = (char*) strchr(buf, '=');
+      if (delim)
+      {
+         *delim++ = '\0';
+
+         key   = buf;
+         value = delim;
+
+         if (!string_is_empty(key) && !string_is_empty(value))
+         {
+            if (string_is_equal(key, "tunnel_addr"))
+               strlcpy(host_room->mitm_address, value,
+                  sizeof(host_room->mitm_address));
+            else if (string_is_equal(key, "tunnel_port"))
+               sscanf(value, "%i", &host_room->mitm_port);
+         }
+      }
+
+      remaining -= (size_t)lnbreak - (size_t)buf;
+      buf = lnbreak;
+   } while (remaining);
+
+   free(buf_start);
+
+done:
+   *finished = true;
+}
+
+static bool netplay_mitm_query(const char *mitm_name)
+{
+   char query[512];
+   net_driver_state_t  *net_st    = &networking_driver_st;
+   struct netplay_room *host_room = &net_st->host_room;
+   const char          *url       = "http://lobby.libretro.com/tunnel";
+   bool                finished   = false;
+
+   if (string_is_empty(mitm_name))
+      return false;
+
+   snprintf(query, sizeof(query), "%s?name=%s", url, mitm_name);
+
+   if (!task_push_http_transfer(query,
```

Comment:
here is a clarification on the changes that were made to task_queue_wait by the author:

the code sends a GET query request to http://lobby.libretro.com/tunnel to translate mitm server names into addresses and ports. We cannot proceed without that info, so task_queue_wait is used in combination with a variable to suspend execution.

He requested me to review this part with him, so I'd appreciate it if me and @jdgleaver can review this particula bit of the PR to ensure nothing goes wrong in the process.

Reply:
Yes, waiting on this task seems like the right thing to do, but the implementation here is not really correct.

At present, this will just wait until all tasks are complete (`netplay_mitm_query_wait` is non-functional in this context)

Either this should be done 'properly', as here for example: https://github.com/libretro/RetroArch/blob/9982452c1821e785f0bbac3af2ff821bc252e913/tasks/task_menu_explore.c#L229-L245

Or if you do want to wait for all tasks, just call `task_queue_wait(NULL, NULL)`...